main/src/ca/uwaterloo/flix/api/lsp/provider/completion/Differ.scala

```diff
@@ -26,20 +26,77 @@ object Differ {
     case None =>
       // Case 1: No old AST. No difference.
       DeltaContext(Nil)
-    case Some(oldAst) => {
+    case Some(oldAst) =>
       // Case 2: We have an oldAst and a newAst. Compute their difference.
 
       // TODO: Add some comments and introduce helper functions.
-      val newDefs = (newAst.defs.keySet -- oldAst.defs.keySet).toList.map(sym => Delta.AddDef(sym, getCurrentTimestamp))
+      val newDefs = findModifiedDefs(oldAst, newAst)
       val newEnums = (newAst.enums.keySet -- oldAst.enums.keySet).toList.map(sym => Delta.AddEnum(sym, getCurrentTimestamp))
 
       DeltaContext(newDefs ++ newEnums)
-    }
+  }
+
+  /**
+    * Find all defs that has been modified.
+    *
+    * This is done by checking if a def in an oldAst is different from that same def in the newAst.
+    * If the def is modified, we add it to the list of deltas.
+    *
+    * @param oldAst the old AST.
+    * @param newAst the new AST.
+    * @return       a List[Delta] consisting of newly modified defs.
+    */
+  private def findModifiedDefs(oldAst: TypedAst.Root, newAst: TypedAst.Root): List[Delta] = {
+    newAst.defs.flatMap {
+      case (sym, newDef) =>
+        // Check if the sym also exists in the oldAst
+        oldAst.defs.get(sym) match {
+          case None =>
+            // It's a newly added def, add it to the deltas
+            Some(Delta.ModifiedDef(sym, getCurrentTimestamp))
+          case Some(oldDef) =>
+            // We have an old def, check if the implementation has changed
+            if (isDefModified(oldDef, newDef)) {
+              // The def has been modified, add it to the deltas
+              Some(Delta.ModifiedDef(sym, getCurrentTimestamp))
+            } else {
+              // The def hasn't changed, don't add it to the deltas.
+              None
+            }
+        }
+    }.toList
+  }
+
+  /**
+    * Compares two defs in terms of change
+    *
+    * @param oldDef the old def.
+    * @param newDef the new def.
+    * @return       true if the def has been modified, false otherwise.
+    */
+  private def isDefModified(oldDef: TypedAst.Def, newDef: TypedAst.Def): Boolean = {
+    !equalLocationSpan(oldDef.impl.exp.loc, newDef.impl.exp.loc)
+  }
+
+  /**
+    * Checks if two locations have equal span.
+    *
+    * This function is different from SourceLocation.equals().
+    * This only checks if the location of a specific node in the AST has the same span as the other.
```

Comment:
I am not sure I understand. Can we discuss in person?

I also now realize that if one function is changed (made longer) would that mark every function afterwards as changed?